service/matching/version_sets.go

```diff
@@ -297,24 +302,34 @@ func lookupVersionSetForPoll(data *persistencespb.VersioningData, caps *commonpb
 	return getSetID(set), nil
 }
 
-// Requires: data is not nil
 func lookupVersionSetForAdd(data *persistencespb.VersioningData, stamp *commonpb.WorkerVersionStamp) (string, error) {
 	var set *persistencespb.CompatibleVersionSet
 	if stamp == nil {
-		// if this is a new workflow, assign it to the latest version.
-		// (if it's an unversioned workflow that has already completed one or more tasks, then
-		// leave it on the unversioned one. that case is handled already before we get here.)
-		setLen := len(data.VersionSets)
+		// If this is a new workflow, assign it to the latest version.
+		// (If it's an unversioned workflow that has already completed one or more tasks, then
+		// leave it on the unversioned one. That case is handled already before we get here.)
+		setLen := len(data.GetVersionSets())
 		if setLen == 0 || data.VersionSets[setLen-1] == nil {
 			return "", errEmptyVersioningData
 		}
 		set = data.VersionSets[setLen-1]
 	} else {
-		// for add, any version in the compatible set maps to the set
+		// For add, any version in the compatible set maps to the set.
+		// Note data may be nil here, findVersion will return -1 then.
 		setIdx, _ := findVersion(data, stamp.BuildId)
 		if setIdx < 0 {
-			// TODO: consider making an ephemeral set so we can match even if replication fails
-			return "", errBuildIDNotFound
+			// A workflow has a build ID set, but we don't know about that build ID. This can
+			// happen in replication scenario: the workflow itself was migrated and we failed
+			// over, but the versioning data hasn't been migrated yet. Instead of rejecting it,
+			// we can guess a set ID based on the build ID. If the build ID was the first in
+			// its set on the other side, then our guess is right and things will work out. If
+			// not, then we'll guess wrong, but when we get the replication event, we'll merge
+			// the sets and use both ids.
+			// TODO: this doesn't really work unless we persist the fact that we've created
```

Comment:
The problem here is:
- we add a task
- guess the set id because we don't recognize the build id
- the task gets spooled
- the tq gets unloaded
- now we get version info that says that build id belongs to another set
- now we load the tq and try to dispatch the task
We've lost the fact that we spooled the task under a guessed set id. And any poll that comes in now will not trigger a load of that set id. So I think we do need to persist this somewhere. On the root we can trigger a mutation. On non-root, it's not clear. Maybe we have to send a message to the root?

Reply:
Oh yeah, that's an issue, we'll need to persist but I'm a bit weary of adding versioning data with incorrect timestamps, that may mess with the merge results and revert user intent.
I'll try to think of the exact scenario where that can happen.

Reply:
could we add it with a fake timestamp far in the past?

Reply:
I'm thinking we might need to store this outside of the version sets to avoid messing with the merge logic. Additionally, it could be nice if a user can query and find out which build IDs have not been replicated so they can manually recover (probably on a namespace level not a task queue level though).

Reply:
discussed on zoom and decided: yes, we'll persist it as a version set. the root can do it directly, others can send an rpc to the root. use a fake zero timestamp. don't make the default (of course). visibility is through metrics and logs for now (though once it's in the version data, it should be easy enough to flag it somehow in describetaskqueue or other rpcs)